src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/segmentation/KernelSegmenter.java

```diff
@@ -0,0 +1,526 @@
+                costsForMergingSegmentPairs.set(indexOfLeftSegmentToMerge, costsForSegmentPairs.get(indexOfLeftSegmentToMerge) - costsForMergedSegmentPairs.get(indexOfLeftSegmentToMerge));
+            }
+
+            //update total segmentation costs and changepoints
+            totalSegmentationCosts.add(0, segments.stream().mapToDouble(s -> s.cost).sum());
+            changepoints.add(0, mergepoint);
+        }
+
+        //find optimal number of changepoints according to penalty function
+        final int effectiveMaxNumChangepoints = Math.min(maxNumChangepoints, changepoints.size());
+        final List<Double> totalSegmentationCostsPlusPenalties = IntStream.range(0, effectiveMaxNumChangepoints + 1).boxed()
+                .map(i -> totalSegmentationCosts.get(i) + changepointPenalties.get(i))
+                .collect(Collectors.toList());
+        final int numChangepointsOptimal = totalSegmentationCostsPlusPenalties.indexOf(Collections.min(totalSegmentationCostsPlusPenalties));
+
+        logger.info(String.format("Found %d changepoints after applying penalties.", numChangepointsOptimal));
+        return changepoints.subList(0, numChangepointsOptimal);
+    }
+
+    /**
+     * Calculates the cost of a segment.  This is defined by Eq. 11 of
+     * <a href="https://hal.inria.fr/hal-01413230/document">https://hal.inria.fr/hal-01413230/document</a>
+     * (except we use the low-rank approximation to the kernel, as described in Sec. 3.2, ibid).
+     * Various recurrence relations are used to calculate costs iteratively.
+     * @param start inclusive start index of segment
+     * @param end   inclusive end index of segment
+     * @param reducedObservationMatrix      N x p matrix of projected observations, where N is the number of data points
+     *                                      and p is the dimension of the low-rank approximation of the kernel matrix;
+     *                                      this is the Z matrix described in the text preceding Eq. 14, ibid
+     * @param kernelApproximationDiagonal   N diagonal terms of the low-rank approximation to the kernel matrix
+     */
+    private static Cost calculateSegmentCost(final int start,
+                                             final int end,
+                                             final RealMatrix reducedObservationMatrix,
+                                             final double[] kernelApproximationDiagonal) {
+        final int N = reducedObservationMatrix.getRowDimension();
+        final int p = reducedObservationMatrix.getColumnDimension();
+
+        //initialize quantities for recurrence
+        double D = kernelApproximationDiagonal[start];
+        final double[] W = Arrays.copyOf(reducedObservationMatrix.getRow(start), p);
+        double V = Arrays.stream(W).map(w -> w * w).sum();
```

Comment:
Might be worth extracting a `MathUtils` for sum of squares of a `double[]`.

Reply:
Where are the boy scouts of yesteryear...